Add drop down unit variant selector

andreievg opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #2156

# 👩🏻‍💻 What does this PR do? 

Add drop down unit variant selector to item list and adjusts quantities based on the selected variant. Also moved sync log datetime to be excluded in refresh date cli, see test description.

# 🧪 How has/should this change been tested? 

Initialise from reference data:
```
cargo run --bin remote_server_cli -- initialise-from-export -n reference1
```
then refresh dates
```
cargo run --bin remote_server_cli -- refresh-dates
```

Check out item view as per screenshot above, try changing unit variants, see stats column reflect selected variant. Also selected variant will be saved, but not in app data (not on refresh, but if navigating to another page through menu), saving userSelected through app data will be added in next task, I added this to epic (btw without user selected, should default to calculated by server, as per mock data for now).

![Screenshot 2023-08-30 at 7 39 09 PM](https://github.com/openmsupply/open-msupply/assets/26194949/a34735c2-af19-4ad2-943f-1d2a77fd5442)

## 💌 Any notes for the reviewer?

The task in epic suggests the memoisation should be done, this was already one through zutand selector (i.e. when new variant is selected through drop down, only the cells for that item should re-render

andreievg commented on client/packages/system/src/Item/Components/ItemVariant/PackUnitSelectCell.tsx:
```diff
@@ -0,0 +1,43 @@
+
+    if (!variantsControll) {
+      // TODO should do default if not units ? (check what happens)
+      return <InnerBasicCell isError={isError} value={getUnitName(rowData)} />;
```
If no config is present, default to unit name in Unit/Pack column

andreievg commented on client/packages/system/src/Item/ListView/ListView.tsx:
```diff
@@ -9,10 +9,15 @@ import {
+  NumUtils,
 } from '@openmsupply-client/common';
-import { useItems, ItemRowFragment } from '../api';
+import { useItems, ItemsWithStatsFragment } from '../api';
```
I added a fragment to reflect the query result, this is still a bit loose in terms of type safety, I think ideally UI will be tightly tied to the data that spawns it, the query (but that would mean removing context).

Sorry this is a little bit out of scope

andreievg commented on client/packages/system/src/Item/api/operations.graphql:
```diff
@@ -171,6 +171,22 @@ query itemStockOnHand(
   }
 }
 
+fragment ItemsWithStats on ItemNode {
```
this is where the fragment is added

andreievg commented on client/packages/system/src/Item/context/useUnitVariant.ts:
```diff
@@ -1,31 +1,40 @@
   items: {
     [itemId: string]: UnitVariantNode;
   };
+  userSelectedVariants: UserSelectedVariants;
```
This structure will be stored in app data and loaded on store change (stored be store)

andreievg commented on client/packages/system/src/Item/context/useUnitVariant.ts:
```diff
@@ -46,33 +55,64 @@ export const useInitUnitStore = () => {
-  // TODO [state.items[itemId], state.userSelected[itemId]]
-  const item = useUnitStore(state => state.items[itemId], isEqual);
-  if (!item) {
+export const useUnitVariant = (
```
this diff is a bit weird, i added type return and then extra selector for useUnitStore

andreievg commented on client/packages/system/src/Item/context/useUnitVariant.ts:
```diff
@@ -46,33 +55,64 @@ export const useInitUnitStore = () => {
 
-  const mostUsedVariant = item.variants.find(
-    ({ id }) => id === item.mostUsedVariantId
-  )?.packSize;
```
previously mostUsedVariant was = packSize (as @mark-prins pointed out in previous review, i missed that)

andreievg commented on server/cli/src/refresh_dates.rs:
```diff
@@ -40,18 +40,6 @@ fn get_timestamp_fields() -> Vec<TableAndFieldName> {
         ("requisition_line", "snapshot_datetime"),
         ("stocktake", "created_datetime"),
         ("stocktake", "finalised_datetime"),
-        ("sync_log", "started_datetime"),
```
Removed because it was messing with date refresh cli (which tries to find the latest date, then difference between that date and now and adjusts all dates by that difference, of course sync log had latest date, which was populated with init from reference data cli, thankfully there was a clear message about this, so just moved to ignore sync log datetimes)

mark-prins reviewed (commented):
I was reflecting on the `Pack/Unit` header and having a mix of numbers (pack size) and text (pack names) in the column. 

<img width="139" alt="image" src="https://github.com/openmsupply/open-msupply/assets/9192912/e13a9497-d9fa-485f-98aa-ab239c93f52c">

I think this will be confusing and wonder instead:
- have the heading just as `Pack`
- if there is a unit name, show as `1 tab` instead of just `tab` (or `100 tabs` etc)
- if there isn't a unit name, show `1 unit` (or `100 units` etc)

that way the list heading is simpler and possibly clearer, and the column content is consistent.
It requires adding an `s` at the end which might not work in other languages, so perhaps we sacrifice that part and show `100 tab`.

mark-prins commented on client/packages/system/src/Item/context/useUnitVariant.ts:
```diff
@@ -1,31 +1,40 @@
+      })),
     setItems: newItems =>
-      set(() => {
+      set(({ userSelectedVariants }) => {
```
some thoughts on the `items` assignment, which I must have missed previously:

I would usually spread first and assign second, so that later items overwrite earlier ones if the id is the same, shouldn't be an issue here, it's just that usually the object creation like this is to patch an object and overwrite entries with new values

```
(acc, item) => ({ ...acc, [item.itemId]: item }),
```

I thought 'surely we've done this before..' and 'why isn't this a helper method?'
Maybe add to `ObjUtils` or `ArrayUtils`?

I also wondered about how efficient the reduce approach is, given that it has to create a new object for each array entry. So I tried this approach:

```
        const items = {};
        newItems.forEach(item => {
          Object.defineProperty(items, item.itemId, { value: item });
        });
```

and ran both options over 5,000 very simple item objects (just an itemId and numeric value) giving this result:

reduce: 1048.97705078125 ms
defineProperty: 0.751953125 ms

```

const items = Array.from(new Array(5000), (_val, idx) => ({itemId: idx, value: idx}));
console.time('reduce');
        const i2 = items.reduce(
          (acc, item) => ({ [item.itemId]: item, ...acc }),
          {}
        );
console.timeEnd('reduce');

console.time('defineProperty');
const i1 = {};
items.forEach(item => {
  Object.defineProperty(i1, item.itemId, { value: item });
});
console.timeEnd('defineProperty');
```

I think it's worth changing! would be good to have as a utility function too in this case, as I expect that I'll try and write a version with reduce next time I need this

andreievg replied:
Thanks Mark !

Added keyBy to utils and used in useVariant: 923e1080b57fac050340fc2557fea023906b0d2b

Benched keyBy just in case (btw thanks for the really cool bench example with console.time):
![Screenshot 2023-09-05 at 12 34 17 PM](https://github.com/openmsupply/open-msupply/assets/26194949/0ba031de-f935-4bd6-918f-071017c654dd)

Looks like lodash uses it's own optimised impleentation
https://github.com/lodash/lodash/blob/2da024c3b4f9947a48517639de7560457cd4ec6c/.internal/arrayReduce.js#L12

mark-prins reviewed (commented):
overall - very slick! nice to use, works well and the code looks great.
I have a possible performance suggestion and a question about column heading / data
Also wondered if this will be applied to the stock list too?

andreievg commented on client/packages/system/src/Item/Components/ItemVariant/PackUnitSelectCell.tsx:
```diff
@@ -0,0 +1,42 @@
+    const { variants, activeVariant, setUserSelectedVariant } =
+      variantsControll;
+
+    console.log(activeVariant);
```
805cd3083e8d1bb97367e86725b6a4e7b8426b01

andreievg commented on client/packages/system/src/Item/Components/ItemVariant/PackUnitSelectCell.tsx:
```diff
@@ -0,0 +1,42 @@
+    getUnitName: (row: T) => string;
+  }) =>
+  ({ isError, rowData }: CellProps<T>): ReactElement => {
+    const { variantsControll } = useUnitVariant(getItemId(rowData));
```
27a1fe6f113863a71296b1f07c6d392782c05e2a

andreievg commented on client/packages/system/src/Item/Components/ItemVariant/PackUnitSelectCell.tsx:
```diff
@@ -0,0 +1,42 @@
+        }}
+        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
+          setUserSelectedVariant(e.target.value);
+        }}
```
cae3e71f7085a2e831a6d443b2e348f86ef7527a

andreievg commented on client/packages/system/src/Item/ListView/ListView.tsx:
```diff
@@ -25,30 +30,33 @@ const ItemListComponent: FC = () => {
+      {
+        key: 'packUnit',
+        label: 'label.pack-unit',
+        align: ColumnAlign.Right,
```
> think there's an exception somewhere, but generally: right align for numbers, left for text

I would argue this is the exception

I think this is better on the right side, since the column is closely linked to the numbers, it helps understand what the numbers mean, even without the packUnit things. I think having it on the right side helps address the other point about KDD, the one about helping user with consistent and clear column names and column placement.

mark-prins replied:
agree - thnx

andreievg commented:
@mark-prins, thanks for review, I think i addressed all of the comments, and made a separate PR for one of them: https://github.com/openmsupply/open-msupply/pull/2177

mark-prins reviewed (approved):
🙏  thanks - looks great!

andreievg commented:
Thanks for review !
